### ISSUE TITLE
Laravel 5.2 make:auth middleware does not work

### ISSUE THREAD

**Issue by Cyrille37** [laravel/framework#12076]
Hi,
I looking at Laravel 5.2 and the easy to use "make:auth" and the multi-auth.

I think there is a bug because in the generated middleware **"App\Http\Middleware\Authenticate" at handle() method call the $guard parameter is always null**.
So Laravel use "default"=>"gard"=>"xyz defined in "auth.php", not the one defined by the $guard property in App\Http\Controllers\Auth\AuthController.

Before the App\Http\Middleware\Authenticate::handle(...) call, **multi-auth** works because the method Illuminate\Foundation\Auth\AuthenticatesUsers::handleUserWasAuthenticated() is well called with the good $this->getGuard() value.

**Comment by Cyrille37** [comment]
info: The debugger show me that Illuminate\Auth\AuthManager::gards is empty ... while terminating successfully Illuminate\Foundation\Auth\AuthenticatesUsers::handleUserWasAuthenticated()

**Comment by Cyrille37** [comment]
After the well logged redirection the Illuminate\Auth\AuthManager is instanciated 2 times and only the default guard "users" appears :

 Illuminate\Auth\\**AuthManager::__construct**
 Illuminate\Auth\AuthManager::guard guards: 
 Illuminate\Auth\AuthManager::resolve name: users
 Illuminate\Auth\\**AuthManager::createSessionDriver name: users**
 Illuminate\Foundation\Auth\AuthenticatesUsers::getGuard guard: admins
 Illuminate\Auth\AuthManager::guard guards: users
 Illuminate\Auth\AuthManager::resolve name: salaries
 Illuminate\Auth\\**AuthManager::createSessionDriver name: admins**
 Illuminate\Foundation\Auth\AuthenticatesUsers::handleUserWasAuthenticated sucess, redirect to "/home"
 App\Http\Middleware\Authenticate::handle guard: 
 Illuminate\Auth\\**AuthManager::__construct**
 Illuminate\Auth\AuthManager::guard guards: 
 Illuminate\Auth\AuthManager::resolve name: users
 Illuminate\Auth\\**AuthManager::createSessionDriver name: users**
 Illuminate\Auth\\**AuthManager::__construct**
 Illuminate\Auth\AuthManager::guard guards: 
 Illuminate\Auth\AuthManager::resolve name: users
 Illuminate\Auth\\**AuthManager::createSessionDriver name: users**
 Illuminate\Auth\AuthManager::guard guards: user

**Comment by rginnow** [comment]
I'm also having an issue with this. I cannot start my application with the built in authentication methods. I have also come to the same conclusion that the guards are always returning null. I've followed the documentation to the T and can't figure out why this happens.

**Comment by ifaniqbal** [comment]
We are having this issue too. I have tried it in 5.1.11 and all work perfectly.

These are some links that discuss this problem:
https://laracasts.com/discuss/channels/laravel/auth-session-killed-after-redirect-laravel-52

http://stackoverflow.com/questions/35074688/laravel-5-2-authattempt-method-does-not-persistently-logs-in-users

**Comment by hiendv** [comment]
Please provide reproduction: routing, controllers, etc..
Did you specify `$guard` on `AuthController`?

**Comment by Cyrille37** [comment]
Hi
I'll post all code if this following does not enough:

- composer new project
- artisan make:auth
- define 2 guards (users & admins) in config/auth.php and 2 that extends Illuminate\Foundation\Auth\User

scenario #1 :
- it works with default guard
- run login = it works

scenario #2 :
- add protected $guard = 'admin'; // (the second guard that is not the default one)
- run login = it does not work: Login successful but not reminded so redirected to "login" not at "home".

**Comment by hiendv** [comment]
I think you there should be another `AuthController` for `admins` guard. That's the way I organize things and it's doing just fine =D

**Comment by rginnow** [comment]
Can you post an example of your second auth controller?

Sent using CloudMagic [https://cloudmagic.com/k/d/mailapp?ct=pi&cv=7.1.27&pv=8.4.1]
On Sun, Jan 31, 2016 at 12:46 PM, laravel/framework
<reply@reply.github.com>
wrote:
I think you there should be another AuthController for admins guard. That's the way I organize things and it's doing just fine =D

—
Reply to this email directly or view it on GitHub
[https://github.com/laravel/framework/issues/12076#issuecomment-177563274] .[https://github.com/notifications/beacon/ABdm4qk4Wlq79tniue9o-BKPe4vz1lpcks5pfj_0gaJpZM4HPcWI.gif]

**Comment by Cyrille37** [comment]
@hiendv: But how to do for make the orther `AuthConroller` use "admins" guard ? I used the class property "$guard" and it works for login, but auth was lost after the redirection ...

**Comment by silver-moon** [comment]
You should also create new middleware like `AuthenticateAdmin` and add to your controller:  
```
    public function handle($request, Closure $next)
    {
        if (Auth::guard('admin')->guest()) {
            if ($request->ajax() || $request->wantsJson()) {
                return response('Unauthorized.', 401);
            } else {
                return redirect()->guest('admin/login');
            }
        }
        return $next($request);
    }
```
Or in your controller, you should pass `admin` guard to `auth` middleware like this:
```
class YourController extends Controller
{
    public function __construct()
    {
        $this->middleware('auth:admin');
    }
}
```
This way laravel will authenticate your user using `admin` guard

**Comment by Cyrille37** [comment]
@silver-moon : Thanks for the tip that I well understand as a right solution.

But what about the `AuthController`'s `$guard` property ? What the meaning if it is not propagated in others Laravel services ?

**Comment by hiendv** [comment]
No need to create a new middleware since Laravel automatically resolve that. All you need is a BaseController for every guard. E.g. there are `FrontendController` and `BackendController` corresponding to `$guard`: `users` and `admins`. Other controllers may extend these controllers for the appreciate guard, include AuthController.

**Comment by Cyrille37** [comment]
@hiendv Yes, I've understand that.

But the problem is that the middleware `Authenticate::handle()` always received an empty `$guard` parameter, so the `if (Auth::guard($guard)->guest()) ...` always use the default guard, nor the Controller's one.

**Comment by silver-moon** [comment]
@Cyrille37  
TL;DR: 
- `$guard` property is to authenticate users when they are logging in.
- `auth` middleware authenticates users from the session.  
You just have to use `$guard` on controller which authenticate login action.

As you can see in `App\Http\Controllers\Auth\AuthController`, there is a `Illuminate\Foundation\Auth\AuthenticatesAndRegistersUsers` trait. It uses `Illuminate\Foundation\Auth\AuthenticatesUsers` trait, look at its `login` method:  
```
   if (Auth::guard($this->getGuard())->attempt($credentials, $request->has('remember'))) {
        return $this->handleUserWasAuthenticated($request, $throttles);
    }
```
and `getGuard` method:
```
   return property_exists($this, 'guard') ? $this->guard : null;
```
You can figure it out immediately that Laravel checks `$guard` property to authenticate `users` only when they are logging in.

**Comment by Cyrille37** [comment]
@hiendv Yes, like I say the authentication process works fine with several guards.

The problem is that the **middleware `Authenticate::handle()`** always received an empty `$guard` parameter, so the `if (Auth::guard($guard)->guest()) ...` **always use the default guar**d, nor the authentication's one.

That way when a well done authentication with the guard that is not the default won't pass the auth middleware... **I'm lost...**

**Comment by Cyrille37** [comment]
Ok, I found that the auth middleware (generated by make:auth) is not ready for multi-guard because it uses the default guard : `if (Auth::guard($guard)->guest())...`

So we need to explicitly pass a gard's name like:
- `if (Auth::guard($guard)->guest())...` for default guard ("users" in my case)
- `if (Auth::guard('admins')->guest())...` for other guard ("admins" in my case)

**Comment by Cyrille37** [comment]
The answer is there : https://larachat.slack.com/archives/laravel5/p1454320847000656 as :
` append it to the middleware argument, e.g. 'middleware' => 'auth:api', to use the api authentication.`

**Comment by rginnow** [comment]
So, I've added the normal `web` guard to the middleware that is already included in my auth.php and i still can't figure out what's wrong. Do I have to create my own guards/middleware?

Laravel 5 has introduced some new concepts to me, especially this middleware stuff, so I apologize for my inexperience with this.

Routes:
```
<?php

/*
|--------------------------------------------------------------------------
| Routes File
|--------------------------------------------------------------------------
|
| Here is where you will register all of the routes in an application.
| It's a breeze. Simply tell Laravel the URIs it should respond to
| and give it the controller to call when that URI is requested.
|
*/

// Route::get('phpinfo', function() {
// 	echo phpinfo();
// });

Route::get('/', function() {
	return redirect('admin/dashboard');
});

Route::group(['middleware' => ['web', 'auth:web'], 'prefix'=>'admin', 'namespace'=>'Admin'], function() {
	Route::get('dashboard', 'AdminController@getDashboard')->name('dashboard');
});

/*
|--------------------------------------------------------------------------
| Application Routes
|--------------------------------------------------------------------------
|
| This route group applies the "web" middleware group to every route
| it contains. The "web" middleware group is defined in your HTTP
| kernel and includes session state, CSRF protection, and more.
|
*/

Route::group(['middleware' => ['web']], function () {
  Route::controller('', 'Auth\AuthController');
});
```

auth.php:
```
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Authentication Defaults
    |--------------------------------------------------------------------------
    |
    | This option controls the default authentication "guard" and password
    | reset options for your application. You may change these defaults
    | as required, but they're a perfect start for most applications.
    |
    */

    'defaults' => [
        'guard' => 'web',
        'passwords' => 'users',
    ],

    /*
    |--------------------------------------------------------------------------
    | Authentication Guards
    |--------------------------------------------------------------------------
    |
    | Next, you may define every authentication guard for your application.
    | Of course, a great default configuration has been defined for you
    | here which uses session storage and the Eloquent user provider.
    |
    | All authentication drivers have a user provider. This defines how the
    | users are actually retrieved out of your database or other storage
    | mechanisms used by this application to persist your user's data.
    |
    | Supported: "session", "token"
    |
    */

    'guards' => [
        'web' => [
            'driver' => 'session',
            'provider' => 'users',
        ],

        'api' => [
            'driver' => 'token',
            'provider' => 'users',
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | User Providers
    |--------------------------------------------------------------------------
    |
    | All authentication drivers have a user provider. This defines how the
    | users are actually retrieved out of your database or other storage
    | mechanisms used by this application to persist your user's data.
    |
    | If you have multiple user tables or models you may configure multiple
    | sources which represent each model / table. These sources may then
    | be assigned to any extra authentication guards you have defined.
    |
    | Supported: "database", "eloquent"
    |
    */

    'providers' => [
        'users' => [
            'driver' => 'eloquent',
            'model' => App\User::class,
        ],

        // 'users' => [
        //     'driver' => 'database',
        //     'table' => 'users',
        // ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Resetting Passwords
    |--------------------------------------------------------------------------
    |
    | Here you may set the options for resetting passwords including the view
    | that is your password reset e-mail. You may also set the name of the
    | table that maintains all of the reset tokens for your application.
    |
    | You may specify multiple password reset configurations if you have more
    | than one user table or model in the application and you want to have
    | separate password reset settings based on the specific user types.
    |
    | The expire time is the number of minutes that the reset token should be
    | considered valid. This security feature keeps tokens short-lived so
    | they have less time to be guessed. You may change this as needed.
    |
    */

    'passwords' => [
        'users' => [
            'provider' => 'users',
            'email' => 'auth.emails.password',
            'table' => 'password_resets',
            'expire' => 60,
        ],
    ],

];
```

AuthController.php:
```
<?php

namespace App\Http\Controllers\Auth;

use App\User;
use Validator;
use App\Http\Controllers\Controller;
use Illuminate\Foundation\Auth\ThrottlesLogins;
use Illuminate\Foundation\Auth\AuthenticatesUsers;

class AuthController extends Controller
{
    /*
    |--------------------------------------------------------------------------
    | Registration & Login Controller
    |--------------------------------------------------------------------------
    |
    | This controller handles the registration of new users, as well as the
    | authentication of existing users. By default, this controller uses
    | a simple trait to add these behaviors. Why don't you explore it?
    |
    */

    use AuthenticatesUsers, ThrottlesLogins;

    /**
     * -------- OVERRIDE TRAIT FUNCTIONS -------------
     */

    protected $username = 'username';
    // protected $guard = 'web';

    /**
     * Show the application login form.
     *
     * @return \Illuminate\Http\Response
     */
    public function getLogin()
    {
      return view('auth.login', ['styles' => '<link href="/css/login.css" rel="stylesheet" type="text/css">']);
    }

		/**
     * -------- END OVERRIDE -------------
     */

    /**
     * Where to redirect users after login / registration.
     *
     * @var string
     */
    protected $redirectTo = '/admin/dashboard';

    /**
     * Create a new authentication controller instance.
     *
     * @return void
     */
    public function __construct()
    {
        $this->middleware('guest', ['except' => 'logout']);
    }
}

```

**Comment by minhchu** [comment]
I think you should post your question on forum https://laracasts.com/discuss/channels/laravel. It's easier for you to get help.

**Comment by GrahamCampbell** [comment]
Or indeed the official laravel forums: http://laravel.io/forum.

**Comment by rginnow** [comment]
OK Thanks, I have posted here: http://laravel.io/forum/02-01-2016-laravel-52-auth-not-working

**Comment by alex-petrea** [comment]
I've had a similar issue once logged in if I were to navigate to any admin area or refresh the page the session would be deleted and redirected back to login. I've fixed it temporary by moving all the grouped middlewares from `web` to global.

Note: it could by that I have custom Auth Middlewares and Services, even though I'm still using Auth::check() to check if user is logged in. 

Note2: I have the same issue on Laravel 5.1 on a different app which started to act weird after many months of working properly and no updates done to it. The problem it's no fixed and I can't seem to find what's causing the problem.

**Comment by rginnow** [comment]
I don't remember how I got around this issue - however, I do know that I had to customize some of the Auth functions inside AuthController.php. I don't remember if indeed it was something to do with the guards or not. It was a lot of debugging though... on both of my projects.

Also, if it helps, I was using GUID and I had to use a UUID helper trait to make it work.
